pkg/user/generator.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2016 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package user
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/coreos/rkt/pkg/group"
+	"github.com/coreos/rkt/pkg/passwd"
+	"github.com/hashicorp/errwrap"
+)
+
+// Generator defines the interface for generating a UID/GID.
+type Generator interface {
+	IDs() (uid int, gid int, err error)
+}
+
+// PickGenerator picks the first non-nil generator and returns it.
+// If no non-nil generator was found, nil is returned.
+func PickGenerator(gs ...Generator) Generator {
+	for _, g := range gs {
+		if g != nil {
+			return g
+		}
+	}
+
+	return nil
+}
+
+type idsFromEtc struct {
+	rootPath string
+	username string
+	group    string
+}
+
+// IDsFromEtc returns a new UID/GID generator by parsing etc/passwd, and etc/group
+// relative from the given rootPath looking for the given username, or group.
+// If username is empty string the etc/passwd lookup will be ommitted.
+// If group is empty string the etc/group lookup will be ommitted.
+func IDsFromEtc(rootPath, username, group string) Generator {
+	return idsFromEtc{
+		rootPath: rootPath,
+		username: username,
+		group:    group,
+	}
+}
+
+func (e idsFromEtc) IDs() (uid int, gid int, err error) {
+	uid, gid = -1, -1
+
+	uid, err = passwd.LookupUidFromFile(
+		e.username,
+		filepath.Join(e.rootPath, "etc/passwd"),
+	)
+
+	if e.username != "" && err != nil {
+		return
+	}
+
+	gid, err = group.LookupGidFromFile(
+		e.group,
+		filepath.Join(e.rootPath, "etc/group"),
+	)
+
+	if e.group != "" && err != nil {
+		return
+	}
+
+	return uid, gid, nil
+}
+
+type idsFromStat struct {
+	path string
+	r    *UidRange
+}
+
+// IDsFromStat returns a new UID/GID generator deriving the UID/GID from file attributes
+// and unshifts the UID/GID if the given range is not nil.
+// If the given id does not start with a slash "/" nil is returned.
+func IDsFromStat(rootPath, file string, r *UidRange) Generator {
+	if strings.HasPrefix(file, "/") {
+		return idsFromStat{filepath.Join(rootPath, file), r}
+	}
+
+	return nil
```

Comment:
Don't do this please. User has to start checking for `nil` values, which is tedious and error-prone and if it receives a nil he won't know the reason, worse it won't be traceable without a proper error.

Instead modify the signature to `(Generator, error)` from `Generator`